### PR TITLE
Eliminate max/min distinct

### LINKED ISSUE #2045: [Optimizer] Eliminate the distinct
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
```sql
explain verbose select max(distinct(c1)) from test;

+-------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type                                             | plan                                                                                                                                        |
+-------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------+
| initial_logical_plan                                  | Projection: #MAX(DISTINCT test.c1)                                                                                                          |
|                                                       |   Aggregate: groupBy=[[]], aggr=[[MAX(DISTINCT #test.c1)]]                                                                                  |
|                                                       |     TableScan: test projection=None                                                                                                         |
| logical_plan after simplify_expressions               | SAME TEXT AS ABOVE                                                                                                                          |
| logical_plan after eliminate_filter                   | SAME TEXT AS ABOVE                                                                                                                          |
| logical_plan after common_sub_expression_eliminate    | SAME TEXT AS ABOVE                                                                                                                          |
| logical_plan after eliminate_limit                    | SAME TEXT AS ABOVE                                                                                                                          |
| logical_plan after projection_push_down               | Projection: #MAX(DISTINCT test.c1)                                                                                                          |
|                                                       |   Aggregate: groupBy=[[]], aggr=[[MAX(DISTINCT #test.c1)]]                                                                                  |
|                                                       |     TableScan: test projection=Some([0])                                                                                                    |
| logical_plan after filter_push_down                   | SAME TEXT AS ABOVE                                                                                                                          |
| logical_plan after limit_push_down                    | SAME TEXT AS ABOVE                                                                                                                          |
| logical_plan after SingleDistinctAggregationToGroupBy | Projection: #MAX(DISTINCT test.c1)                                                                                                          |
|                                                       |   Projection: #MAX(alias1) AS MAX(DISTINCT test.c1)                                                                                         |
|                                                       |     Aggregate: groupBy=[[]], aggr=[[MAX(#alias1)]]                                                                                          |
|                                                       |       Aggregate: groupBy=[[#test.c1 AS alias1]], aggr=[[]]                                                                                  |
|                                                       |         TableScan: test projection=Some([0]) 
```

**Describe the solution you'd like**
 I think `max/min` don't need the rewrite, but need to eliminate the distinct.

**Describe alternatives you've considered**
`SingleDistinctAggregationToGroupBy` rule make the plan more complex.

**Additional context**
None

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #2045 .

 # Rationale for this change

related: #1315

# What changes are included in this PR?

Eliminate distinct in the aggregate instead of rewrite

Now the target is rewrite from

```sql
| initial_logical_plan | Projection: #Max(DISTINCT test.c1)                     
|                      |   Aggregate: groupBy=[[]], aggr=[[Max(DISTINCT #test.c1)]]                      
|                      |     TableScan: test projection=None
```
to
```sql
| initial_logical_plan | Projection: #Max(DISTINCT test.c1)
|                      |   Projection: #Max(#test.c1) AS Max(DISTINCT test.c1)                   
|                      |     Aggregate: groupBy=[[]], aggr=[[Max(#test.c1)]]                
|                      |       TableScan: test projection=None
```
instead of 

```sql
| logical_plan  | Projection: #Max(DISTINCT test.c1)          
|               |   Projection: #Max(alias1) AS Max(DISTINCT test.c1)     
|               |     Aggregate: groupBy=[[]], aggr=[[Max(#alias1)]]        
|               |       Aggregate: groupBy=[[#test.c1 AS alias1]], aggr=[[]]                        
|               |       Aggregate: groupBy=[[#test.c1 AS alias1]], aggr=[[]]  
|               |         TableScan: test projection=Some([0])
```

# Are there any user-facing changes?
None

### CODE REVIEW

**Comment by jackwener** [comment]
I encountered some problems during development. Ask for some help. ❤🙌

key point is how to rewrite the `shema` and `expr` of `alias_projection` and `rewrited_aggregation`. It's will make the entire plan suitable.

Now my code is with bugs. And I can't  rewrite the `shema` and `expr`  successful to fix it.

**Comment by jackwener** [comment]
@Dandandan @alamb @ic4y @houqp PTAL, Thanks! ❤

**Comment by jackwener** [comment]
I found that draft PR also trigger CI

File: datafusion/src/optimizer/single_distinct_to_groupby.rs [hunk 1]
```diff
@@ -193,7 +285,7 @@ impl OptimizerRule for SingleDistinctToGroupBy {
         optimize(plan)
     }
     fn name(&self) -> &str {
-        "SingleDistinctAggregationToGroupBy"
+        "single_distinct_aggregation_to_groupBy"
```
Comment (Dandandan) [inline datafusion/src/optimizer/single_distinct_to_groupby.rs:288]:
```suggestion
        "single_distinct_aggregation_to_group_by"
```

File: datafusion/src/optimizer/single_distinct_to_groupby.rs [hunk 2]
```diff
@@ -62,6 +63,10 @@ fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
             group_expr,
         }) => {
             if is_single_distinct_agg(plan) {
+                if is_max_or_min(plan) {
```
Comment (alamb) [inline datafusion/src/optimizer/single_distinct_to_groupby.rs:66]: I don't think this optimization is limited to plans with a single distinct aggregate (the way it is for count). 

In other words, I think it is possible to *always* rewrite `min(distinct x)` to `min(x) as "min(distinct x)"` and `max(distinct x)` to `max(x) as "max(distinct x)"`

Comment (jackwener) [reply datafusion/src/optimizer/single_distinct_to_groupby.rs:66]: Great point, thanks!❤

**Comment by jackwener** [comment]
I found that problem occur when  call `ExprSchema data_type()`, it will be error.

Because `index_of_column_by_name` can't get `test.c1`  from `max(test.c1)`

**Comment by jackwener** [comment]
Now I rewrite the logical plan. and `println!` log in terminal.

I think it's ok. but there will be error.

```sql
Projection: #MAX(#test.c1) AS MAX(DISTINCT test.c1) 
DFSchema { fields: [DFField { qualifier: None, field: Field { name: "MAX(DISTINCT test.c1)", data_type: Float32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None } }], metadata: {} }
Expr[MAX(#test.c1) AS MAX(DISTINCT test.c1)]

	Aggregate: groupBy=[[]], aggr=[[MAX(#test.c1)]]
	DFSchema { fields: [DFField { qualifier: None, field: Field { name: "MAX(test.c1)", data_type: Float32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None } }], metadata: {} }
	Expr[MAX(#test.c1)]

		TableScan: test projection=Some([0])
		DFSchema { fields: [DFField { qualifier: Some("test"), field: Field { name: "c1", data_type: Float32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: None } }], metadata: {} }
```

**Pull request closed without merging**